docs/sql-data-sources-parquet.md

```diff
@@ -252,6 +252,72 @@ REFRESH TABLE my_table;
+sc.hadoopConfiguration.set("parquet.crypto.factory.class" ,
+                   "org.apache.parquet.crypto.keytools.PropertiesDrivenCryptoFactory")
+
+// Write encrypted dataframe files. 
+// Column "square" will be protected with master key "keyA".
+// Parquet file footers will be protected with master key "keyB"
+squaresDF.write.
+   option("parquet.encryption.column.keys" , "keyA:square").
+   option("parquet.encryption.footer.key" , "keyB").
+parquet("/path/to/table.parquet.encrypted")
+
+// Read encrypted dataframe files
+val df2 = spark.read.parquet("/path/to/table.parquet.encrypted")
+
+{% endhighlight %}
+
+</div>
+
+
+#### KMS Client
+
+The InMemoryKMS class is provided only for illustration and simple demonstration of Parquet encryption functionality. **It should not be used in a real deployment**. The master encryption keys must be kept and managed in a production-grade KMS system, deployed in user's organization. Rollout of Spark with Parquet encryption requires implementation of a client class for the KMS server. Parquet provides a plug-in [interface](https://github.com/apache/parquet-mr/blob/apache-parquet-1.12.0/parquet-hadoop/src/main/java/org/apache/parquet/crypto/keytools/KmsClient.java) for development of such classes,
+
+<div data-lang="java"  markdown="1">
+{% highlight java %}
+
+public interface KmsClient {
+  // Wraps a key - encrypts it with the master key.
+  public String wrapKey(byte[] keyBytes, String masterKeyIdentifier);
+
+  // Decrypts (unwraps) a key with the master key. 
+  public byte[] unwrapKey(String wrappedKey, String masterKeyIdentifier);
+
+  // Use of initialization parameters is optional.
+  public void initialize(Configuration configuration, String kmsInstanceID, 
+                         String kmsInstanceURL, String accessToken);
+}
+
+{% endhighlight %}
+
+</div>
+
+An [example](https://github.com/apache/parquet-mr/blob/apache-parquet-1.12.0/parquet-hadoop/src/test/java/org/apache/parquet/crypto/keytools/samples/VaultClient.java) of such class for an open source [KMS](https://www.vaultproject.io/api/secret/transit) can be found in parquet-mr repository. The production KMS client should be designed in cooperation with organization's security administrators, and built by developers with an experience in access control management. Once such class is created, it can be passed to applications via the `parquet.encryption.kms.client.class` Hadoop parameter and leveraged by general Spark users as shown in the dataframe write/read sample above.
+
+Note: By default, Parquet implements "double envelope encryption" mode, that minimizes the interaction of Spark executors with a KMS server. In this mode, the DEKs are encrypted with "key encryption keys" (KEKs, randomly generated by Parquet). The KEKs are encrypted with MEKs; the result and the KEK itself are cached in Spark executor memory. Users interested in a regular envelope encryption, can switch it on via configuration.
```

Comment:
Please mention the exact configuration name here: `can switch it on via configuration.` -> `can switch it on via xxx`.